program/src/processor.rs

```diff
@@ -640,44 +695,108 @@ pub fn process_withdraw_inactive_stake(
         // withdraw from these accounts, so we should not observe a decrease in
         // balance.
         msg!(
             "Observed balance of {} is less than tracked balance of {}.",
             observed_total,
-            validator.entry.stake_accounts_balance
+            validator.entry.effective_stake_balance()
         );
         msg!("This should not happen, aborting ...");
         return Err(LidoError::ValidatorBalanceDecreased.into());
     }
 
     // We tracked in `stake_accounts_balance` what we put in there ourselves, so
     // the excess is a donation by some joker.
-    let donation = (observed_total - validator.entry.stake_accounts_balance)
+    let donation = (observed_total - validator.entry.effective_stake_balance())
         .expect("Does not underflow because observed_total >= stake_accounts_balance.");
     msg!("{} in donations observed.", donation);
 
     // Store the new total. If we withdrew any inactive stake back to the
     // reserve, that is now no longer part of the stake accounts, so subtract
     // that.
-    validator.entry.stake_accounts_balance =
-        (observed_total - excess_removed).expect("Does not underflow, because excess <= total.");
+    validator.entry.stake_accounts_balance = (validator.entry.stake_accounts_balance
+        - excess_removed)
+        .expect("Does not underflow, because excess <= total.");
+
+    // Try to withdraw from unstake accounts.
+    let mut unstake_removed = Lamports(0);
+    observed_total = Lamports(0);
+    for seed in &validator.entry.unstake_seeds {
+        let (unstake_account_address, _bump_seed) =
+            validator.find_unstake_account_address(program_id, accounts.lido.key, seed);
+
+        let unstake_account = compare_stake_account(
+            &mut stake_accounts,
+            n_unstake_accounts as usize,
+            n_stake_accounts as usize,
+            seed,
+            &unstake_account_address,
+            &accounts,
+            StakeType::Unstake,
+        )?;
+        let account_balance = Lamports(**unstake_account.lamports.borrow());
+        msg!(
+            "Unstake account at seed {} ({}) contains {}.",
+            seed,
+            unstake_account.key,
+            account_balance
+        );
+
+        // If validator's stake is at the beginning, try to withdraw.
+        if validator.entry.unstake_seeds.begin == seed {
+            let unstake_removed_here = withdraw_inactive_sol(
+                &WithdrawExcessOpts {
+                    accounts: &accounts,
+                    clock: &clock,
+                    rent: &rent,
+                    stake_history: &stake_history,
+                    stake_account: unstake_account,
+                },
+                seed,
+                lido.stake_authority_bump_seed,
+                StakeType::Unstake,
+            )?;
+            // If withdrew everything, bump the seed
+            if account_balance == unstake_removed_here {
```

Comment:
And then, when we always withdraw 100% or nothing, this can become an assertion instead of a check.